### ISSUE TITLE
[Implementation] Different DOIs for different article versions

### ISSUE THREAD

**Issue by bozana** [pkp/pkp-lib#10553]
The DOI backend allows assignment of different DOIs to different article versions. This option currently only exists in OPS. For the project OSS ORE an assignment of different DOIs to different article versions is needed also in OJS.

The decisions made:
All JAV will get a DOI, so that preprints can get a DOI.
The minor versions in the same version stage and same version major will have one and the same DOI. This requires changes in the current code.

Only current publication is displayed in submission listing on the DOI management page (as it was till now). The link 'View all' opens a modal where the last minor version for all major versions will be listed. This is slightly different as till now (in OPS, where all versions are listed and all versions get a DOI). 
We cannot display all versions i.e. also all minor versions, because each version can be edited, which could lead in wrongly editing minor versions (so that it could happen that they have different DOIs) i.e. we would need to consider this logic also in the edit function in UI (so that when editing a minor version all sibling minor version would be edited in the same way). This is a bigger task (for me) and I think does not need to be implemented right now.

Tasks to do:
- [x] Add setting for a journal to define if each article version will get a new DOI
- [x] Consider it on the DOI management page
- [x] Consider it for display on the article landing page. If a minor version does not have a DOI, we check if there is a sibling minor version, to display the DOI from there, because all minor versions (of a submission and with the same version stage and major) have the same DOI. This is considered also for chapters, but not for galleys, publication formats and files, because it is not possible to find out the sibling minor version of them. Currently we do not display DOIs for galley and files, and DOIs for these object types are not much used.
- ~~[ ] Consider it in native import export plugin~~ -- The DOIs will be exported for publications (and other objects) that have a DOI in the DB. Differently to the DOI display on the landing page, if a minor version do not have a DOI, it will not be searched for a DOI of a sibling minor version for export/import. 

PR:

s. comment regarding development notes: https://github.com/pkp/pkp-lib/issues/10553#issuecomment-3045723046

- https://github.com/pkp/pkp-lib/pull/11625
- https://github.com/pkp/ui-library/pull/690
- https://github.com/pkp/ojs/pull/4982
- https://github.com/pkp/omp/pull/2091
- https://github.com/pkp/ops/pull/1076

Fix marking DOIs stale:
- https://github.com/pkp/pkp-lib/pull/11835
- https://github.com/pkp/ojs/pull/5099
- https://github.com/pkp/omp/pull/2124
- https://github.com/pkp/ops/pull/1099

Another fix marking DOIs stale:
- https://github.com/pkp/pkp-lib/pull/11845

Fix typo in two OMP locale keys:
- https://github.com/pkp/omp/pull/2137

**Comment by bozana** [comment]
A = one DOI for all versions
B = one DOI for a version

Here I will write notes during the implementation:

publication()::version()
A: the DOI will be copied
B: **changed**: if major version the DOI will be set to null. If minor version the DOI will be copied.

publication()::publish()
A: mark only that publication's DOIs stale, if the publication is current publication (to update DOI metadata, to point to the new current publication) --> **changed**, earlier all submission's DOIs were marked stale.
B: If it is a major version, mark all submission's versions' DOIs stale (to update relationships between different DOIs). If it is a minor version and the current publication, mark this verion's DOIs stale. 

PublicationPublished event (triggered in publication()::publish(), where also DOIs are marked stale)
-> VersionDois listener
**Changed**: Instead of calling submission()->createDois() (that considers only current publication), publication()->createDois will be called, so that this publication can get the DOI, if it does not have any yet.
This is needed because the published version is not necessarily the current publication and more.

publication()::unpublish()
A: If the unpublished publication was the current publication, mark the unpublished publication's DOIs stale (so that the DOIs can be re-registered and DOI point to the new current publication). **Changed**: earlier the submission status (is published) was considered: if the submission status was not published any more, all submission's DOIs wrere marked stale. Now, we do not consider submission status any more, because submission status should be decoupled from publication statuses. 
B: If it was a major version or a current minor version, mark the the unpublished publication's DOIs stale. 

submission()->createDois()
Called in 
DecisionAdded event -> AssignDOIs listener
PKPDoiController::assignSubmissionsDois()
A: and B: current publication will get a DOI
It could be that some earlier publication will not have a DOI (if DOI management started later)

markSubmissionsStale() (done manually, e.g. if metadata changes)
A: and B: Applied if submission is published and if current publication DOI status is either registered or submitted -- i.e. only current publication is considered. Then all submission's DOIs are marked stale.
(On next deposit:
A: current publication DOI metadata will be updated
B: all DOIs will be updated)


When using the Bulk function Assign DOI we assign a DOI only to the current publication and its sub-elements. 

Still Open:
- See how publication workflow and statuses develop, then eventually adapt  DecisionAdded event -> AssignDOIs listener 


Edge cases:
- When enabling B after A:
Some major versions could have the same DOI. This can be problem if already registered submission gets a new publication or is re-registered, because we register all versions for B. The problem can be decreased if we register versions from the less 'mature' to most 'mature' one -- this way, the DOI for the same versions will at least have the most actual metadata.
- When enabling A after B:
All next versions would have the same DOI as the current publication. This is less a problem because for registration we only consider the current version.
On article landing page we would see only one DOI for all publication, which is not true.
The registered DOIs should not be edited -- it would not change all relevant, same DOIs, but only all major versions in that stage and major version. 
Worse scenario, because existing DOIs for versions cannot be seen in UI.
- When enabling B later (no DOIs have existed yet):
Minor versions can not get a DOI, only the last one. Older minor versions are not visible in the UI, only the last one or if current publication.
Situation when version 1.0 is current publication and there is also version 1.1, then: When assigning a DOI, version 1.0 will get a DOI because it is the current version. When version 1.1 is published, it will get a new DOI, because it existing earlier, before DOIs were assigned, so that DOI could not be copied from the version 1.0 to 1.1. The same applies for galleys.
We do not have information of what other galley a galley is version of -- this connection could be lost if galley file changes. Thus it is probably best to assign a new DOI to version 1.1, even it is a minor version, so that is can be registered as a new DOI, that would contain the correct galleys relationships, and information about the galleys DOIs. 
If we would keep the same DOI also for version 1.1, and if galley files have changed, then the galleys from 1.1 version would get new DOIs. On the registration the DOI would point to the new galleys, wile the DOIs for the old galleys from the version 1.0 would stay unconsidered, not connected to any DOI.
- When enabling A later (no DOIs have existed yet):
Only the current publication gets a DOI. This if not a problem at all.
o

**Comment by bozana** [comment]
Hi @ewhanson, could you please review the PRs above?
Thanks a lot!

**Comment by ewhanson** [comment]
Thanks @bozana, I've finished a review of the PRs and it looks really good. Mostly just a few questions about comments/clarity and one about locale keys. I don't think there's need for another review again, at least from me. 👍

**Comment by bozana** [comment]
Hi @Tribunal33 

This issue, that implements assignment of different DOIs to different versions, has been merged. There is still some UX/UI improvements to be decided on and then made later, s. https://github.com/pkp/pkp-lib/issues/11819.

How generally DOI assignment should behave in workflow, e.g. on publication version, publish and unpublish, see this comment above: https://github.com/pkp/pkp-lib/issues/10553#issuecomment-3045723046. And, here I will try to explain the requirements for this new feature (to assign different DOIs to different versions) as good as possible for you to be able to test:

- There is new DOI setting (Settings > Distribution > DOIs > Setup) in OJS and OMP for assigning different DOIs to different versions -- in OPS it was already there. In OJS and OMP it is per default set to "No", in OPS to "Yes".
- If the setting is "No", then one and the same DOI is assigned to all versions. This was so until now in OJS and OMP. This applies also to all sub-elements, like galleys, chapters, etc. 
- If the setting is "Yes", then all major versions should get a different DOI. All minor versions, of the same version stage and version major, have one and the same DOI. This applies also to all sub-elements, like galleys, chapters, etc. 
- It is important to note that on _publication versioning_, the metadata are copied from one to the new version, also the DOIs in case of a minor version. A minor version can only be created from the last mature stage, s. https://github.com/pkp/pkp-lib/issues/11781 -- this is still not implemented but important to consider when testing DOIs assignment. In case of a major version, the DOIs are not copied to the new version.
Also consider the above mentioned edge case, when there are already versions in the system before the DOI assignment is set up. So it is maybe best to test it all on new submissions (and not the one that come with our usual dataset).
- Note that in the UI the current publication is displayed on the main DOI management page listing. Only if there are other published versions, the possibility to "View all" is displayed.
- When using the Bulk function "Assign DOI" we assign a DOI only to the current publication and its sub-elements.
- If a publication does not have a DOI, e.g. a major publication, a DOI will be assigned to it on publication publishing.
- On a publication/version landing page the DOIs for that version is displayed. If a minor version does not have a DOI, the DOI from its sibling minor version (that has the same DOI) should be displayed.
- Currently we do not display DOIs for galleys (in OJS and OPS), and publication formats and files (in OMP).

(The DOI registration will come later, then it could be easier to test it all fully. I think now only the correct DOI assignment and their display on the landing pages is relevant.)

Thanks a lot!

**Comment by Tribunal33** [comment]
@bozana I think for a limited scope this ticket can pass as I think the basic scenarios are passing.  I have a few issues that I can create sub-issues around so that we don't introduce too much scope creep to this ticket.

- [ ] Currently there is a problem with the stages before copyediting not updating their status to Published which won't show the submission on the DOI management page. 

<img width="1225" height="483" alt="Image" src="https://github.com/user-attachments/assets/d674736b-8279-4f10-9a02-3e958799926d" />

- [ ] The setting for only Allowing DOI assignment only after copyediting stage that seems to assign a DOI no matter what the stage on the Front-office article view

<img width="859" height="359" alt="Image" src="https://github.com/user-attachments/assets/f6719660-364a-4324-a6cd-f107a11b85a6" />

- [ ] Bulk actions not sure how it is supposed to work when there are multiple empty DOIs from various versions and you Bulk Action > Assign DOIs it isn't adding anything.

<img width="1312" height="471" alt="Image" src="https://github.com/user-attachments/assets/096e0985-6e8f-4c1d-a9c5-fcde21de8a6d" />

**Comment by Tribunal33** [comment]
@bozana if those above issue are not in scope then you can close this ticket once the following issue is resolved.  OMP wording isn't spelt correctly.  Do we have any requirements for what the new setting content should be?

<img width="727" height="223" alt="Image" src="https://github.com/user-attachments/assets/5ef46f19-a846-46a7-9314-4502be6f3605" />

Here were the testing scenarios for this feature
1. Test New Setting section is displayed
2. Test Automatic DOI Assignment setting option
3. Test Automatic DOI Assignment setting option only after copyedit stage still functions
4. Test Creating Major Version change creates a new DOI
5. Test Creating 2 Major version changes creates a new unique DOI
6. Test Creating a Minor version does not create a new DOI
7. Test Creating a new major version from lower heirarchy creates a new DOI but isn't displayed
8. Test View All Link opens up a new blade and displays all the various DOIs that were created
9. Test View all link DOIs are correclty matching the Major versions are unique and minor versions are same as Major
10. Test front-office DOI links are referencing the most recent version
11. Test front-office DOI links fallback to correct highest priority most recent version when most recent is unpublished
12. Test DOI Managment list shows highest published version when most recent is unpublished
13. Test manually assigning DOI to minor version and then publishing a minor verison update.
14. Test Article submission wth 20+ versions
15. OMP specific Test DOI Assigned to Chapters

**Comment by bozana** [comment]
Hi @Tribunal33, thanks a lot!
The first scenario should get a separate issue.
The second - when DOI should be assigned on copyediting: I think this works so: the DOI is assigned once the article reaches copyediting or production stage, but also and latest when it is published. So I think this is ok. And that's why the first scenario is strange/bug.
The third scenario is noted here, together with other UI issues: https://github.com/pkp/pkp-lib/issues/11819, under Assign DOIs. Devika suggested to have a button close to every empty DOI field that would generate the DOI for that field. The Bulk assignment should stay as it is - only current version gets a DOI.

**Comment by Tribunal33** [comment]
Thanks @bozana.  I have created a new bug ticket for the first scenario #11886.  I have created a task for the second scenario #11887.  I will leave scenario 3 for now as it may be handled in #11819 as you indicated.  

The remaining issue is the superficial one that I didn't create an bug for and that is the OMP content fix.  That can be handled in the 11886 bug ticket with a minor tweak.  Once it is fixed we can close this ticket. 

I have updated the status of the various projects to back into `in progress`

**Comment by bozana** [comment]
Fixed the typo, thus closing...
